README.md
PBQ
===

[![PyPI version](https://badge.fury.io/py/pbq.svg)](https://badge.fury.io/py/pbq)

[![Build Status](https://travis-ci.org/amirdor/pbq.svg?branch=master)](https://travis-ci.org/amirdor/pbq)

[![Documentation Status](https://readthedocs.org/projects/pbq/badge/?version=latest)](https://pbq.readthedocs.io/en/latest/?badge=latest)

python BiqQuery driver for easy access



Installing
----------

To start using this package run:

``` bash
$ pip install pbq
```

For development use, and local testing run:

``` bash
$ python setup.py install
```



Dependency
----------

**bq Command-Line Tool**

Install bq cli -- the package is running bq cli in the background

``` bash
https://cloud.google.com/sdk/docs/
```

after installation on a new terminal write:

``` bash
$ bq init
$ gcloud auth application-default login
```

if you don't want to install bq cli you can run the package with the
service account key

all you need to do it to define an environment variable like this:

``` bash
$ export GOOGLE_APPLICATION_CREDENTIALS="/home/user/Downloads/[FILE_NAME].json"
```

Usage
=====

Import this package to your project and you are ready to go

``` python
from pbq import PBQ
from pbq import Query
```



PBQ
---

This class can have generic functions for big query manipulation and
save queries to tables.

### some usages:

#### Download table to CSV

``` python
from pbq import PBQ
from pbq import Query

# read sql file and format the query from file to string without parameters
query = Query.read_file('queries/query.sql')

# init the query builder
pbq = PBQ(query)

#run the query and get csv
pbq.to_csv('file.csv', save_query=True, **{'table': 'table', 'dataset': 'dataset'})
```

when setting save\_query=True it means that you want to save your query
to a table, and you need to send dictionary with the table name and
dataset name.

Download table to pandas DataFrame

``` python
from pbq import PBQ
from pbq import Query

# read sql file and format the query from file to string without parameters
query = Query.read_file('queries/query.sql')

# init the query builder
pbq = PBQ(query)

#run the query and get dataframe
pbq.to_dataframe()
```


#### Save query to a table

``` python
from pbq import PBQ
from pbq import Query

# read sql file and format the query from file to string without parameters
query = Query('select * from table')

# init the query builder
pbq = PBQ(query)
pbq.save_to_table('table_name', 'dataset_name', 'project_name', replace=True, partition='20190910')
```

When setting replace=True it will override the table as long as the
table is not partitioned, if partitioned it will overwrite the
partition.

partition format is `YYYYMMDD`



#### upload DataFrame to a table

``` python
import pandas as pd
from pbq import Query, PBQ
df = pd.DataFrame()
PBQ.save_dataframe_to_table(df, 'table', 'dataset', 'project_id', partition='20191013', replace=False)
```



Query
-----

This class will format your query, check validation and return the price
of the query

### some usages:

#### generate the query object with a simple query

``` python
from pbq import Query
query = Query("select * from table")
```

#### get query price

``` python
from pbq import Query
query = Query("select * from table")
print("the query price:", query.price)
# the query price: 0.312
```

#### validate query

``` python
from pbq import Query
query = Query("select * from table")
if not query.validate():
    raise RuntimeError("table not valid")
```

#### query with parameters

``` python
from pbq import Query
query = Query("select * from table where user_id={user_id}", parameters={'user_id': 123})
print(query.query)
# select * from table where user_id=123
```

#### read the query from file with parameters

``` python
from pbq import Query
query = Query.read_file('file_path.sql', parameters={'user_id':123})
print(query.query)
# select * from table where user_id=123
```


-   Free software: MIT license
-   Documentation: <https://pbq.readthedocs.io>.



HISTORY.md
History
=======

0.1.0 (2019-10-03)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for pbq."""


tests/test_pbq.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pbq` package."""


import unittest
from unittest import mock

from pbq import PBQ
from pbq import Query


class TestPBQ(unittest.TestCase):
    """Tests for `pbq` package."""

    @mock.patch('google.cloud.bigquery.Client')
    @mock.patch('google.cloud.bigquery_storage_v1beta1.BigQueryStorageClient')
    def test_arguments(self, mock_Client, mock_BigQueryStorageClient):
        query_str = "select * from users"
        query = Query(query_str)
        pbq = PBQ(query)
        self.assertEqual(pbq.query, query_str)


tests/test_query.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pbq` package."""


import unittest
from unittest import mock

from pbq import Query
import os
import json


class TestQuery(unittest.TestCase):
    """Tests for `pbq` package."""

    @mock.patch('google.cloud.bigquery.Client')
    def test_query_format(self, mock_Client):
        cwd = os.getcwd()
        query_file = '{}/tests/files/sample_query.sql'.format(cwd)
        query = Query.read_file(query_file)
        q_file = open(query_file, 'r')
        _query = q_file.read()
        self.assertEqual(query.query, _query)

    @mock.patch('google.cloud.bigquery.Client')
    def test_query_format_with_variables(self, mock_Client):
        cwd = os.getcwd()
        with open('{}/tests/files/init.json'.format(cwd)) as json_file:
            data = json.load(json_file)
        query = Query.read_file('{}/tests/files/variable_query.sql'.format(cwd), parameters=data)
        q_file = open('{}/tests/files/sample_query.sql'.format(cwd), 'r')
        _query = q_file.read()
        self.assertEqual(query.query, _query)

    @mock.patch('google.cloud.bigquery.Client')
    def test_query_format_with_variables_as_dict(self, mock_Client):
        cwd = os.getcwd()
        query = Query.read_file('{}/tests/files/variable_query.sql'.format(cwd), parameters={'until_day': 1})
        q_file = open('{}/tests/files/sample_query.sql'.format(cwd), 'r')
        _query = q_file.read()
        self.assertEqual(query.query, _query)

    @mock.patch('google.cloud.bigquery.Client')
    def test_query_format_with_missing_values(self, mock_Client):
        cwd = os.getcwd()
        with self.assertRaises(ValueError):
            Query.read_file('{}/tests/files/variable_query.sql'.format(cwd))


pbq/pbq.py
# -*- coding: utf-8 -*-

"""Main module."""
from google.cloud import bigquery
from pbq.query import Query
from google.cloud import bigquery_storage_v1beta1
from google.cloud.exceptions import NotFound
import pandas as pd
import datetime


class PBQ(object):
    """
    bigquery driver using the google official API

    Attributes
    ------
    query : str
        the query
    query_obj : Query
        pbq.Query object
    client : Client
        the client object for bigquery
    bqstorage_client : BigQueryStorageClient
        the google storage client object

    Methods
    ------
    to_dataframe(save_query=False, **params)
        return the query results as data frame

    to_csv(filename, sep=',', save_query=False, **params)
        save the query results to a csv file

    save_to_table(table, dataset, project=None, replace=True, partition=None)
        save query to table

    table_details(table, dataset, project)
        get the information about the table


    Static Methods
    ------
    save_file_to_table(filename, table, dataset, project, file_format=bigquery.SourceFormat.CSV, max_bad_records=0,
                           replace=True, partition=None)
        save file to table, it can be partitioned and it can append to existing table.
        the supported formats are CSV or PARQUET

    save_dataframe_to_table(df: pd.DataFrame, table, dataset, project, max_bad_records=0, replace=True,
                                partition=None)
        same as save file just with pandas dataframe

    table_exists(client: bigquery.Client, table_ref: bigquery.table.TableReference)
        check if table exists - if True - table exists else not exists


    Examples
    ------
    getting query to dataframe

    >>> from pbq import Query, PBQ
    >>> query = Query("select * from table")

    >>> print("the query price:", query.price)

    >>> if not query.validate():
    >>>     raise RuntimeError("table not valid")

    >>> pbq = PBQ(query)
    >>> pbq.to_dataframe()

    saving query to csv

    >>> from pbq import Query, PBQ
    >>> query = Query("select * from table")
    >>> pbq = PBQ(query)
    >>> pbq.to_csv()

    saving dataframe to table

    >>> import pandas as pd
    >>> from pbq import Query, PBQ
    >>> df = pd.DataFrame()

    >>> PBQ.save_dataframe_to_table(df, 'table', 'dataset', 'project_id', partition='20191013', replace=False)
    """

    def __init__(self, query: Query):
        """
        bigquery driver using the google official API
        :param query: Query object


        """
        self.query = query.query
        self.query_obj = query
        self.client = bigquery.Client()
        self.bqstorage_client = bigquery_storage_v1beta1.BigQueryStorageClient()

    def to_dataframe(self, save_query=False, **params):
        """
        return the query results as data frame

        in order to save the query to a table as well as getting the dataframe, send a dict as params with:
        - table
        - dataset
        it will save to the same project

        :param save_query: boolean
            if to save the query to a table also

        :param params: dict
            when `save_query` flag is on you need to give the relevant params

        :return: pd.DataFrame
         the query results
        """
        job_config = bigquery.QueryJobConfig()
        if save_query:
            table_ref = self.client.dataset(params['dataset']).table(params['table'])
            job_config.destination = table_ref

        query_job = self.client.query(query=self.query, job_config=job_config)
        query_job_res = query_job.result()
        df = query_job_res.to_dataframe(bqstorage_client=self.bqstorage_client)
        return df

    def to_csv(self, filename, sep=',', save_query=False, **params):
        """
        save the query results to a csv file

        in order to save the query to a table as well as getting the dataframe, send a dict as params with:
        - table
        - dataset
        it will save to the same project

        :param filename: str
            with the path to save the file

        :param sep: str
            separator to the csv file

        :param save_query: boolean
            if to save the query to a table also

        :param params: dict
            when `save_query` flag is on you need to give the relevant params
        """
        df = self.to_dataframe(save_query, **params)
        df.to_csv(filename, sep=sep, index=False)

    def save_to_table(self, table, dataset, project=None, replace=True, partition=None):
        """
        save query to table

        :param table: str
            table name

        :param dataset: str
            data set name

        :param project: str
            project name

        :param replace: boolean
            if set as true -  it will replace the table, else append to table (default: True)

        :param partition: str
            partition format DDMMYYY (default: None)
        """
        job_config = bigquery.QueryJobConfig()
        # Set the destination table
        client = self.client
        if partition:
            table = '{0}${1}'.format(table, partition)
        table_ref = client.dataset(dataset).table(table.split('$')[0])

        exists_ok = PBQ._writing_disposition(job_config, replace)

        if project:
            table_ref = client.dataset(dataset, project=project).table(table)

        PBQ._create_table(client, exists_ok, partition, replace, table_ref)

        job_config.destination = table_ref
        query_job = client.query(self.query, job_config=job_config)
        query_job.result()
        print('Query results loaded to table {}'.format(table_ref.path))

    @staticmethod
    def _writing_disposition(job_config: bigquery.QueryJobConfig, replace):
        exists_ok = False
        if replace:
            exists_ok = True
            job_config.write_disposition = bigquery.WriteDisposition.WRITE_TRUNCATE
        else:
            job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
        return exists_ok

    @staticmethod
    def _create_table(client: bigquery.Client, exists_ok, partition, replace, table_ref):
        if (partition and not PBQ.table_exists(client, table_ref)) or (not partition and replace):
            bq_table = bigquery.Table(table_ref)
            if partition:
                time_partitioning = bigquery.TimePartitioning()
                bq_table.time_partitioning = time_partitioning
            client.create_table(bq_table, exists_ok=exists_ok)

    @staticmethod
    def save_file_to_table(filename, table, dataset, project, file_format=bigquery.SourceFormat.CSV, max_bad_records=0,
                           replace=True, partition=None):
        """
        save file to table, it can be partitioned and it can append to existing table.
        the supported formats are CSV or PARQUET

        :param filename: str
            with the path to save the file

        :param table: str
            table name

        :param dataset: str
            data set name

        :param project: str
            project name

        :param file_format: str
            possible file format (CSV, PARQUET) (default: CSV)

        :param max_bad_records: int
            number of bad records allowed in file (default: 0)

        :param replace: boolean
            if set as trueit will replace the table, else append to table (default: True)

        :param partition: str
            partition format DDMMYYY (default: None)
        """
        client = bigquery.Client(project=project)

        dataset_ref = client.dataset(dataset)
        if partition:
            table = '{0}${1}'.format(table, partition)
        table_ref = dataset_ref.table(table.split('$')[0])
        job_config = bigquery.LoadJobConfig()
        job_config.max_bad_records = max_bad_records
        job_config.source_format = file_format
        exists_ok = PBQ._writing_disposition(job_config, replace)

        if file_format == bigquery.SourceFormat.CSV:
            job_config.skip_leading_rows = 1
        job_config.autodetect = True

        PBQ._create_table(client, exists_ok, partition, replace, table_ref)

        with open(filename, "rb") as source_file:
            job = client.load_table_from_file(source_file, table_ref, job_config=job_config)

        job.result()  # Waits for table load to complete.
        print("Loaded {} rows into {}:{}.".format(job.output_rows, dataset, table))

    @staticmethod
    def save_dataframe_to_table(df: pd.DataFrame, table, dataset, project, max_bad_records=0, replace=True,
                                partition=None):
        """
        save pd.DataFrame object to table

        :param df: pd.DataFrame
            the dataframe you want to save

        :param table: str
            table name

        :param dataset: str
            data set name

        :param project: str
            project name

        :param max_bad_records: int
            number of bad records allowed in file (default: 0)

        :param replace: boolean
            if set as true -  it will replace the table, else append to table (default: True)

        :param partition: str
            partition format DDMMYYY (default: None)
        """
        now = datetime.datetime.now()
        random_string = '{}'.format(now.strftime('%y%m%d%H%M%S'))
        input_path = "/tmp/tmp-{}.parquet".format(random_string)
        PBQ._save_df_to_parquet(df, input_path)
        PBQ.save_file_to_table(input_path, table, dataset, project, file_format=bigquery.SourceFormat.PARQUET,
                               max_bad_records=max_bad_records, replace=replace, partition=partition)

    @staticmethod
    def _save_df_to_parquet(df, input_path, index=False):
        df.columns = ["{}".format(col) for col in df.columns]
        df.to_parquet(input_path, index=index)

    def table_details(self, table, dataset, project):
        """
        return a dict object with some details about the table

        :param table: str
            table name

        :param dataset: str
            data set name

        :param project: str
            project name

        :return: dict
            with some table information like, last_modified_time, num_bytes, num_rows, and creation_time
        """
        dataset_ref = self.client.dataset(dataset, project=project)
        table_ref = dataset_ref.table(table)
        table = self.client.get_table(table_ref)

        res = {'last_modified_time': table.modified, 'num_bytes': table.num_bytes, 'num_rows': table.num_rows,
               'creation_time': table.created}
        return res

    @staticmethod
    def table_exists(client: bigquery.Client, table_ref: bigquery.table.TableReference):
        """
        check if table exists - if True - table exists else not exists

        :param client: bigquery.Client object

        :param table_ref: bigquery.table.TableReference object
            with the table name and dataset

        :return: boolean
            True if table exists
            False if table not exists
        """
        try:
            table = client.get_table(table_ref)
            if table:
                return True
        except NotFound as error:
            return False


pbq/query.py
from google.cloud import bigquery
import json
import requests
from functools import lru_cache

TERA_IN_BYTES = 1024 * 1024 * 1024 * 1024.0
PRICE_PER_TERA = 5.0


class Query(object):
    """this class control the query, validate, read sql files and return the price.

    Attributes
    ------
    query : str
        the query
    client : Client
        the client object for bigquery


    Methods
    ------
    price()
        return the query price

    validate()
        validate that the query can run without errors


    Static Methods
    ------
    read_file(query_file, parameters=None)
        parse query from file, the query can be with parameters inside the file


    Examples
    ------
    generate Query object with simple query

    >>> from pbq import Query
    >>> query = Query("select * from table")

    get query price

    >>> from pbq import Query
    >>> query = Query("select * from table")
    >>> print("the query price:", query.price)
    # the query price: 0.312

    validate query

    >>> from pbq import Query
    >>> query = Query("select * from table")
    >>> if not query.validate():
    >>>     raise RuntimeError("table not valid")

    query with parameters

    >>> from pbq import Query
    >>> query = Query("select * from table where user_id={user_id}", parameters={'user_id': 123})
    >>> print(query.query)
    # select * from table where user_id=123

    read query from file with parameters

    >>> from pbq import Query
    >>> query = Query.read_file('file_path.sql', parameters={'user_id':123})
    >>> print(query.query)
    # select * from table where user_id=123

    """

    def __init__(self, query, parameters=None):
        """
        :param query: str
            the query

        :param parameters: dict
            key value parameters to change values dynamically inside the query
        """
        self.query = query.replace('"', '\'')
        self._parameters = parameters if parameters else dict()
        self.client = bigquery.Client()
        self._format()

    def _format(self):
        from string import Formatter

        # logger.info("building your stunning query")
        names = [fn for _, fn, _, _ in Formatter().parse(self.query) if fn is not None]

        if len(names) == 0:
            return

        self._validate_params(names)

        self.query = self.query.format_map(self._parameters)
        # logger.info("finish building your query")

    def _validate_params(self, names):
        missing_keys = set(names) - set(self._parameters.keys())
        if len(missing_keys) != 0:
            # logger.error(
            #     "not all the params in the query are set in the config file. missing params: {}".format(missing_keys))
            raise ValueError("not all the params in the query are set in the config file. \n missing params: {}".format(
                missing_keys))

    @staticmethod
    def _js_r(filename):
        with open(filename) as f_in:
            return json.load(f_in)

    @property
    @lru_cache(1)
    def price(self):
        """check the cost of the query

        :return: float
            the price of the query
        """
        res = self._init_query_command()
        if res is None:
            raise RuntimeError("Query is not valid, please fix your query first")
        price = int(res.total_bytes_billed) / TERA_IN_BYTES * PRICE_PER_TERA
        return round(price, 4)

    @lru_cache(1)
    def validate(self):
        """validate the query

        :return: Boolean
            True if the query is runnable

        :raise: RuntimeError
            on query error
        """
        res = self._init_query_command()
        if res is None:
            raise RuntimeError("Query is not valid, please fix your query first")
        return True

    @staticmethod
    def read_file(query_file, parameters=None):
        """parse query from file

        :param query_file: str
            path to the query file

        :param parameters: dict
            key value parameters to change values dynamically inside the query

        :return: Query object
        """
        q_file = open(query_file, 'r')
        _query = q_file.read()
        return Query(_query, parameters)

    @lru_cache(1)
    def _init_query_command(self):
        _query = self.query
        job_config = bigquery.QueryJobConfig()
        job_config.dry_run = True
        job_config.use_query_cache = False
        try:
            query_job = self.client.query(query=_query, job_config=job_config, )  # API request
        except requests.exceptions.HTTPError:
            query_job = None
        return query_job
